### PR TITLE
Generate ZipcodeRates from Weams data

### PR DESCRIPTION
**Pull request by devin-mccurdy**
https://github.com/department-of-veterans-affairs/vets.gov-team/issues/18857

### CODE REVIEW

**Review by zurbergram — APPROVED** [review]
LGTM

**Comment by devin-mccurdy** [comment]
@department-of-veterans-affairs/backend-review-group Could I get a code review on this at your earliest convenience. Thank you!

File: app/models/institution_builder.rb
Comment (jholton) [inline app/models/institution_builder.rb]: Can we get a spec for this in `institution_builder_spec.rb`?

**Review by sonntag-adam — APPROVED** [review]

File: app/models/institution_builder.rb
Comment (dmccurdybah) [reply app/models/institution_builder.rb]: @jholton - spec'd!

File: app/controllers/v0/zipcode_rates_controller.rb
Comment (annaswims) [inline app/controllers/v0/zipcode_rates_controller.rb]: It would be more performant if you wrote it as:  then you could use a composite index. 
`resource = ZipcodeRate.where(version: @version[:number], zip_code: params[:id]).order(:mha_rate).first`

File: db/migrate/20190626110700_add_columns_to_zipcode_rates.rb
Comment (annaswims) [inline db/migrate/20190626110700_add_columns_to_zipcode_rates.rb]: I'd suggest making a composite index with 
`add_index :zipcode_rates, :[zip_code, :version] , using: :btree`

File: app/controllers/v0/zipcode_rates_controller.rb
Comment (annaswims) [reply app/controllers/v0/zipcode_rates_controller.rb]: I take that back.  They both end up with the same sql, though I do think the suggested version is clearer to read.

Comment (dmccurdybah) [reply app/controllers/v0/zipcode_rates_controller.rb]: Can do!

File: app/models/zipcode_rate.rb
Comment (annaswims) [inline app/models/zipcode_rate.rb]: this scope is no longer used anywhere

File: db/migrate/20190626110700_add_columns_to_zipcode_rates.rb
Comment (annaswims) [inline db/migrate/20190626110700_add_columns_to_zipcode_rates.rb]: The colon should be inside the brackets.

**Review by annaswims — DISMISSED** [review]

File: db/migrate/20190626110700_add_columns_to_zipcode_rates.rb
Comment (jholton) [inline db/migrate/20190626110700_add_columns_to_zipcode_rates.rb]: Make sure this runs locally.  Pretty sure there's a typo there.  Should be `[:zip_code, :version]`

Comment (dmccurdybah) [reply db/migrate/20190626110700_add_columns_to_zipcode_rates.rb]: Oof, thanks!

File: db/migrate/20190626110700_add_columns_to_zipcode_rates.rb
Comment (jholton) [inline db/migrate/20190626110700_add_columns_to_zipcode_rates.rb]: Can we swap the order of the columns in this index?  
```suggestion
    add_index :zipcode_rates, [:zip_code, :version] , using: :btree
    add_index :zipcode_rates, [:version, :zip_code] , using: :btree

```
I'm making an assumption that we'll only care about a `zip_code` for a given version.  That is, we're probably never going to look for records for a specific `zip_code` regardless of it's `version`, right?

Changing the order will improve efficiency when just filtering records on `version` too.

File: lib/csv_helper/exporter.rb
Comment (jholton) [inline lib/csv_helper/exporter.rb]: This `false_to_nil` conversion is unnecessary.
Both `false` and `nil` respond to `.present?` with `false`

File: db/migrate/20190626110700_add_columns_to_zipcode_rates.rb
Comment (dmccurdybah) [reply db/migrate/20190626110700_add_columns_to_zipcode_rates.rb]: That's correct.

Can do!

File: app/models/zipcode_rate.rb
Comment (jholton) [inline app/models/zipcode_rate.rb]: I was looking back at the [history](https://github.com/department-of-veterans-affairs/gibct-data-service/commit/fe82564eef46f7320cfedcf3b85c75376820fb46) of this file.
It looks like this class has never been updated by a CSV uploaded via the dashboard (and thus has never been "exported" either).
As far as I can tell, the table has only been populated by a seed job.
That seed job utilized this `CSV_CONVERTER_INFO` hash.

If from now on this table will be populated by data in the WEAM spreadsheet, this can be an empty class:
```
class ZipcodeRate < ActiveRecord::Base
end
```
* we don't need to include `CsvHelper` since we're not loading or exporting.
* we don't need this `CSV_CONVERTER_INFO` hash at all for the same reason.
* the `validates :version, presence: true` is never used either since we're doing a direct SQL insert.

File: app/models/institution_builder.rb
Comment (jholton) [inline app/models/institution_builder.rb]: What about the other `ZipcodeRate` attributes?
`mha_name`
`mha_code`
I was searching around the `vets-website` codebase, and I'm not sure that `mha_code` is used anywhere, but `mha_name` definitely _is_ used.

**Review by jholton — CHANGES_REQUESTED** [review]

File: app/models/institution_builder.rb
Comment (dmccurdybah) [reply app/models/institution_builder.rb]: They're not currently in the WEAMS data, and we have approval from the VA to remove `mha_name` from vets-website.

Comment (jholton) [reply app/models/institution_builder.rb]: Are they going to be added to the WEAMS data?  If not, should we drop those columns from the table, and remove references to them from the codebase (e.g. `ZipcodeRateSerializer`)?

Comment (dmccurdybah) [reply app/models/institution_builder.rb]: Unknown.  We're going to hold off making any changes until later because getting this in as-is is very high priority for them.

File: lib/csv_helper/exporter.rb
Comment (jholton) [inline lib/csv_helper/exporter.rb]: It looks like all of these changes in this file revolve around getting the value of a method call into the `Institutions` export.
I propose a simpler solution:
Revert this entire file, and just change each usage of `record[k]` to `record.public_send(k)`.  I count 3 places.
And you can remove `non_hash: true` up in the `Institution` model.

It should be save to use `public_send` since we are only `send`ing methods hard coded in the models (that is, not `send`ing user input)

Let me know if I'm misunderstanding something.

**Review by jholton — CHANGES_REQUESTED** [review]

File: lib/csv_helper/exporter.rb
Comment (dmccurdybah) [reply lib/csv_helper/exporter.rb]: No, that makes sense and works.  Thanks!

**Review by jholton — APPROVED** [review]
LGTM, with a caveat:
We will need to run a SQL update on `zipcode_rates` to populate the new `version` column for the existing data.  Otherwise `zipcode_rates_controller#show` will start throwing Exceptions for every request.

**Comment by devin-mccurdy** [comment]
We're expecting that and are planning on generating a new preview version in staging.  I believe that we've set the expectation that this is something we will have to coordinate with the VA when we want this to be moved to production.

**Pull request merged**
